Registry/RepositoryRegistry.php

```diff
@@ -0,0 +1,147 @@
+<?php
+
+/*
+ * This file is part of the Symfony CMF package.
+ *
+ * (c) 2011-2016 Symfony CMF
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Cmf\Bundle\ResourceBundle\Registry;
+
+use Symfony\Cmf\Component\Resource\RepositoryFactoryInterface;
+use Symfony\Cmf\Component\Resource\RepositoryRegistryInterface;
+use Puli\Repository\Api\ResourceRepository;
+
+/**
+ * @author Daniel Leech <daniel@dantleech.com>
+ */
+class RepositoryRegistry implements RepositoryRegistryInterface
+{
+    private $factories = [];
+    private $configurations = [];
+    private $instances = [];
+    private $typeMap = [];
+
+    public function __construct(
+        array $configurations
+    ) {
+        $this->configurations = $configurations;
+    }
+
+    /**
+     * Add a repository factory.
+     *
+     * @param string $name
+     * @param RepositoryFactoryInterface
+     */
+    public function addFactory($name, RepositoryFactoryInterface $factory)
+    {
+        if (isset($this->factories[$name])) {
+            throw new \RuntimeException(sprintf(
+                'A factory named "%s" has already been registered',
+                $name
+            ));
+        }
+
+        $this->factories[$name] = $factory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function get($instanceName)
+    {
+        if (!isset($this->instances[$instanceName])) {
+            $this->instances[$instanceName] = $this->createRepositoryInstance($instanceName);
+        }
+
+        return $this->instances[$instanceName];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getRepositoryAlias(ResourceRepository $repository)
+    {
+        foreach ($this->instances as $alias => $repositoryInstance) {
+            if ($repositoryInstance === $repository) {
+                return $alias;
+            }
+        }
+
+        throw new \InvalidArgumentException(sprintf(
+            'Could not determine alias of repository of class "%s", it was not created by this registry.',
+            get_class($repository)
+        ));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getRepositoryType(ResourceRepository $repository)
+    {
+        $repositoryClass = get_class($repository);
+        if (!isset($this->typeMap[$repositoryClass])) {
+            throw new \InvalidArgumentException(sprintf(
+                'No repository has been instantiated of class "%s", cannot determine the type, known types: "%s".',
+                $repositoryClass,
+                implode('", "', $this->typeMap)
+            ));
+        }
+
+        return $this->typeMap[$repositoryClass];
+    }
+
+    private function createRepositoryInstance($instanceName)
```

Comment:
You mean create the instances in the DI extension and register them here? I suppose that could be a better solution. Will have a look.